### PR TITLE
[UNR-1997] 4.23.1 Upgrade

### PR DESCRIPTION
**Pull request by joshuahuburn**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
4.23.1! 
These are the changes necessary to support the engine upgrade to 4.23.1. Engine changes can be previewed here: https://github.com/improbableio/UnrealEngine/pull/279
There are additional macros to ensure compatibility with 4.22 as the next release (0.8.0) will be 4.22 and 4.23 compatible. 

@m-samiec @mironec @improbable-valentyn @Vatyx

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 1]
```diff
@@ -34,41 +34,53 @@ DECLARE_CYCLE_STAT(TEXT("ReplicateActor"), STAT_SpatialActorChannelReplicateActo
 DECLARE_CYCLE_STAT(TEXT("UpdateSpatialPosition"), STAT_SpatialActorChannelUpdateSpatialPosition, STATGROUP_SpatialNet);
 DECLARE_CYCLE_STAT(TEXT("ReplicateSubobject"), STAT_SpatialActorChannelReplicateSubobject, STATGROUP_SpatialNet);
 
+#if ENGINE_MINOR_VERSION <= 22
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: If this is only intended for local use, move it inside the namespace

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 2]
```diff
@@ -902,7 +952,13 @@ FObjectReplicator* USpatialActorChannel::PreReceiveSpatialUpdate(UObject* Target
 
 	FObjectReplicator& Replicator = FindOrCreateReplicator(TargetObject).Get();
 	TargetObject->PreNetReceive();
+#if ENGINE_MINOR_VERSION <= 22
 	Replicator.RepLayout->InitShadowData(Replicator.RepState->StaticBuffer, TargetObject->GetClass(), (uint8*)TargetObject);
+#else
+	//TODO: This is basically a workaround where we do not construct a proper ShadowData state, so we just reset it here - this also probably leaks memory if the shadow data has arrays
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Can we construct a proper ShadowData state instead?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialFastArrayNetSerialize.cpp [hunk 3]
```diff
@@ -76,5 +77,37 @@ void SpatialFastArrayNetSerializeCB::NetSerializeStruct(UScriptStruct* Struct, F
 		RepLayout_SerializePropertiesForStruct(*RepLayout, Ar, PackageMap, reinterpret_cast<uint8*>(Data), bHasUnmapped);
 	}
 }
+#else
+//TODO: Probably copy more from FNetSerializeCB::NetSerializeStruct
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialFastArrayNetSerialize.cpp]: What's the action item here?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 4]
```diff
@@ -64,6 +64,12 @@ USpatialNetDriver::USpatialNetDriver(const FObjectInitializer& ObjectInitializer
 	, NextRPCIndex(0)
 	, TimeWhenPositionLastUpdated(0.f)
 {
+#if ENGINE_MINOR_VERSION >= 23
+	// Due to changes in 4.23, we now use an outdated flow in ComponentReader::ApplySchemaObject
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Is this intended to be fixed before merging?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 5]
```diff
@@ -342,7 +348,14 @@ void USpatialNetDriver::CreateServerSpatialOSNetConnection()
 	check(NetConnection);
 
 	ISocketSubsystem* SocketSubsystem = GetSocketSubsystem();
+	// This is just a fake address so that Unreal doesn't ensure-crash on disconnecting from SpatialOS
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Only in 4.23?

File: SpatialGDK/Source/SpatialGDK/SpatialGDK.Build.cs [hunk 6]
```diff
@@ -48,40 +48,44 @@ public SpatialGDK(ReadOnlyTargetRules Target) : base(Target)
         string SharedLibSuffix = "";
         bool bAddDelayLoad = false;
 
-        switch (Target.Platform)
+        if (Target.Platform == UnrealTargetPlatform.Win32 || Target.Platform == UnrealTargetPlatform.Win64)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/SpatialGDK.Build.cs]: Why the change from switch to if?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 5 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: The crash only happens in 4.23, but the change should not affect the correctness of 4.22, which is why I didn't wrap it in a guard.

File: SpatialGDK/Source/SpatialGDK/SpatialGDK.Build.cs (hunk 6 above)
Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/SpatialGDK.Build.cs]: Target.Platform was changed from a string (? I think) to a class or another type that is no longer able to be switched on.

**Comment by oblm** [comment]
Please add the following to the `CHANGELOG.md`

> Unreal Engine `4.23.1` is now supported. You can find the `4.23.1` version of our engine fork [here](https://github.com/improbableio/UnrealEngine/tree/4.23-SpatialOSUnrealGDK).

**Comment by oblm** [comment]
* How was this tested?
* Considering the areas of risk that you discovered during this work, how do you propose I test it further?
* Please make changes to `docs-preview` to list this as our new supported engine version.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 7]
```diff
@@ -221,13 +233,23 @@ void USpatialActorChannel::UpdateShadowData()
 	}
 
 	// Refresh shadow data when crossing over servers to prevent stale/out-of-date data.
+#if ENGINE_MINOR_VERSION <= 22
 	ActorReplicator->RepLayout->InitShadowData(ActorReplicator->ChangelistMgr->GetRepChangelistState()->StaticBuffer, Actor->GetClass(), (uint8*)Actor);
+#else
+	ActorReplicator->ChangelistMgr->GetRepChangelistState()->StaticBuffer.Buffer.Empty();
+	ActorReplicator->RepLayout->InitRepStateStaticBuffer(ActorReplicator->ChangelistMgr->GetRepChangelistState()->StaticBuffer, (const uint8*)Actor);
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Update to c++ style casts.

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialFastArrayNetSerialize.h [hunk 8]
```diff
@@ -21,6 +21,15 @@ class SpatialFastArrayNetSerializeCB : public INetSerializeCB
 	{ }
 
 	virtual void NetSerializeStruct(UScriptStruct* Struct, FBitArchive& Ar, UPackageMap* PackageMap, void* Data, bool& bHasUnmapped);
+#if ENGINE_MINOR_VERSION >= 23
+	virtual void NetSerializeStruct(FNetDeltaSerializeInfo& Params);
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialFastArrayNetSerialize.h]: Probs should checkf here too

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialFastArrayNetSerialize.h [hunk 9]
```diff
@@ -21,6 +21,15 @@ class SpatialFastArrayNetSerializeCB : public INetSerializeCB
...
+
+	//TODO (UNR-2371): Look at whether we need to implement these - added check(false)s here to guarantee no entry
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialFastArrayNetSerialize.h]: Update comments

File: SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp [hunk 10]
```diff
@@ -255,11 +255,16 @@ TSharedPtr<FUnrealType> CreateUnrealTypeInfo(UStruct* Type, uint32 ParentChecksu
 	// Based on inspection in InitFromObjectClass, the RepLayout will always replicate object properties using NetGUIDs, regardless of
 	// ownership. However, the rep layout will recurse into structs and allocate rep handles for their properties, unless the condition
 	// "Struct->StructFlags & STRUCT_NetSerializeNative" is true. In this case, the entire struct is replicated as a whole.
+#if ENGINE_MINOR_VERSION <= 22
 	FRepLayout RepLayout;
 	RepLayout.InitFromObjectClass(Class);
-	for (int CmdIndex = 0; CmdIndex < RepLayout.Cmds.Num(); ++CmdIndex)
+	FRepLayout* RepLayoutPtr = &RepLayout;
+#else
+	TSharedPtr<FRepLayout> RepLayoutPtr = FRepLayout::CreateFromClass(Class, nullptr/*ServerConnection*/, ECreateRepLayoutFlags::None);
```
Comment (joshuahuburn) [inline SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp]: Let's take a look at how we're doing this.

Comment (mironec) [reply SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/TypeStructure.cpp]: Reduced the amount of changes here.

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialFastArrayNetSerialize.h [hunk 11]
```diff
@@ -12,15 +12,24 @@ class USpatialNetDriver;
...
 class SpatialFastArrayNetSerializeCB : public INetSerializeCB
 {
 public:
 	SpatialFastArrayNetSerializeCB(USpatialNetDriver* InNetDriver)
 		: NetDriver(InNetDriver)
 	{ }
-
 	virtual void NetSerializeStruct(UScriptStruct* Struct, FBitArchive& Ar, UPackageMap* PackageMap, void* Data, bool& bHasUnmapped);
+#if ENGINE_MINOR_VERSION >= 23
+	//TODO: UNR-2371 - Look at whether we need to implement these and implement 'NetSerializeStruct(FNetDeltaSerializeInfo& Params)'.
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialFastArrayNetSerialize.h]: Please add `override` to each of these

File: SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialConditionMapFilter.h [hunk 12]
```diff
@@ -31,7 +31,11 @@ class FSpatialConditionMapFilter
 #endif
 
 		// Build a ConditionMap. This code is taken directly from FRepLayout::RebuildConditionalProperties
+#if ENGINE_MINOR_VERSION <= 22
 		static_assert(COND_Max == 14, "We are expecting 14 rep conditions"); // Guard in case more are added.
+#else
+		static_assert(COND_Max == 16, "We are expecting 16 rep conditions"); // Guard in case more are added.
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialConditionMapFilter.h]: 2 conditions have been added in 4.23, but we've only added one more entry?

File: SpatialGDK/Source/SpatialGDK/SpatialGDK.Build.cs (hunk 6 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/SpatialGDK.Build.cs]: 🤢

File: SpatialGDK/Source/SpatialGDK/Public/Utils/RepLayoutUtils.h [hunk 13]
```diff
@@ -192,7 +192,14 @@ inline TArray<UFunction*> GetClassRPCFunctions(const UClass* Class)
 	}
 
 	// When using multiple EventGraphs in blueprints, the functions could be iterated in different order, so just sort them alphabetically.
-	RelevantClassFunctions.Sort([](const UFunction& A, const UFunction& B) { return A.GetFName() < B.GetFName(); });
+	RelevantClassFunctions.Sort([](const UFunction& A, const UFunction& B)
+	{
+#if ENGINE_MINOR_VERSION <= 22
+		return A.GetFName() < B.GetFName();
+#else
+		return FNameLexicalLess()(A.GetFName(), B.GetFName());
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/RepLayoutUtils.h]: Was the FName < operator removed in 4.23?

Comment (mironec) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/RepLayoutUtils.h]: It was deprecated, since it does expensive lexical sorting, but that is exactly the functionality we want. There is also FNameFastLess() which is faster, but is not stable between runs (so not usable for us).

File: SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialConditionMapFilter.h (hunk 12 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDK/Public/Interop/SpatialConditionMapFilter.h]: This is a fun one. The old `COND_Max` used to be 14, they removed 14 and then made the new condition `COND_Never=15`. They then made the new `COND_Max=16. So basically they only added one new condition and they incremented twice :p

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialFastArrayNetSerialize.h (hunk 11 above)
Comment (joshuahuburn) [reply SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialFastArrayNetSerialize.h]: Done
